### DISCUSSION TITLE
How to turn off doc-in-minibuffer but keep calling syntax?

### DISCUSSION THREAD

**Discussion by jdtsmith** [joaotavora/eglot#734] [Q&A] [answered]
I'd like to change eglot to use a key combo to display documentation of the thing at point (in another buffer), but preserve auto-display of calling syntax.  With pylsp I get multiple lines of minibuffer info for booleans variables, for example.  

In images:

Good:

![image](https://user-images.githubusercontent.com/93749/132094834-f692d04c-5dba-4d9f-afff-af28679570b3.png)

Less good:

![image](https://user-images.githubusercontent.com/93749/132094853-d924cffa-10a6-461f-b276-5da570e9c866.png)

**Comment by joaotavora** [comment, marked as answer]
Try `M-x eldoc`, bind it to a key, and play around with `eldoc-echo-area...` variables.

**Comment by jdtsmith** [comment]
Thanks.  Is it possible to configure syntax (like my first image) and docs (like the 2nd) separately?  Ideally I'd like the former to appear automatically in the echo area, and the latter to be summoned by key stroke.

**Reply by joaotavora** [reply]
I think it is. Did you find anything interesting in the variables I suggested? Did you try my suggestion of M-x eldoc?

**Reply by jdtsmith** [reply]
I found that `M-x eldoc-mode` turns off eldoc, so both syntax and docs are no longer inserted automatically.  And then you can summon `eldoc` whenever.    This bring up an eldoc buffer into which both syntax and docs will go.  Since docs and syntax are treated the same way here — either both showing or both not — I haven't found how to make syntax auto-show in the echo area, and docs only on key press.  

The variables you mention simply control how auto-displayed eldoc text is truncated or redirected to the eldoc buffer when showing.

**Reply by joaotavora** [reply]
I told you about `M-x eldoc`, not `M-x eldoc-mode`.  As to what is "signatures" and what is "doc", that seems to change with the LSP server. 

> The variables you mention simply control how auto-displayed eldoc text is truncated or redirected to the eldoc buffer when showing.

They are usually sufficient for most users here: "signatures" come first and occupy the first line, which escapes truncation. The rest goes into the buffer.  To mix and match specific eldoc backends with specific eldoc outlets requires more `eldoc.el` work (which I'm doing as well), but it's not supported yet.

**Reply by jdtsmith** [reply]
Thanks.  A reasonably good compromise is `(setq eldoc-echo-area-use-multiline-p nil)`.  I do wish the eldoc buffer when summoned would shrink to fit the buffer. Can probably advise `eldoc--format-doc-buffer` to do so.

**Reply by joaotavora** [reply]
> Thanks. A reasonably good compromise is (setq eldoc-echo-area-use-multiline-p nil). 

Yes, I had a feeling that your request wasn't the first in this regard, and I worked in `eldoc.el` to satisfy similar requests.

I've noticed that most of the work I intended to do in `eldoc.el` is already done.  Look at `eldoc-display-functions` in the latest Eldoc (which should be installed as a package or in Emacs 28).

What's missing is a way for backends in `eldoc-documentation-functions` to hint to each element of `eldoc-display-function` that the documentation items they produce should or shouldn't be handled by that `e-d-f` element. Maybe a small amount of work needs still to be done in `eldoc.el` and/or Eglot.

>  I do wish the eldoc buffer when summoned would shrink to fit the buffer. 

That is a fairly common request that I wouldn't mind accommodating.  Please open a bug with `M-x report-emacs-bug` and cc me in the recipients.

**Comment by kootenpv** [comment]
@jdtsmith perfect question

@joaotavora perfect answer, this worked: `(setq eldoc-echo-area-use-multiline-p nil)`
